onnxruntime/core/providers/cpu/math/einsum_utils.cc

```diff
@@ -0,0 +1,724 @@
+  for (size_t i = 0; i < lo.size(); ++i) {
+    output_dims.push_back(left_dims[lo[i]]);
+  }
+
+  if (!is_final_pair) {
+    for (size_t i = 0; i < reduce_dims.size(); ++i) {
+      output_dims.push_back(1);  // reduced dimensions will have a value 1 in it
+    }
+  }
+
+  for (size_t i = 0; i < ro.size(); ++i) {
+    output_dims.push_back(right_dims[ro[i]]);
+  }
+
+  std::vector<int64_t> current_subscript_order;
+
+  // Calculate output permutation
+  // After the MatMul op, the because the two operands have been permutated,
+  // the output is permutated as well with respect to the original ordering of the axes.
+  // The permutated order will be the dims in: [lro, lo, reduced_dims, ro]
+  // Hence invert the permutation by a permutation that puts the axes in the same ordering
+  std::vector<size_t> output_permutation;
+  if (!is_final_pair) {  // If this is not the final pair, we need to permutate the result to match the pre-fixed order for the next iteration
+    output_permutation.resize(lro.size() + lo.size() + reduce_dims.size() + ro.size(), 0);
+    size_t iter = 0;
+    for (size_t i = 0; i < lro.size(); ++i) {
+      output_permutation[lro[i]] = iter++;
+    }
+    for (size_t i = 0; i < lo.size(); ++i) {
+      output_permutation[lo[i]] = iter++;
+    }
+    for (size_t i = 0; i < reduce_dims.size(); ++i) {
+      output_permutation[reduce_dims[i]] = iter++;
+    }
+    for (size_t i = 0; i < ro.size(); ++i) {
+      output_permutation[ro[i]] = iter++;
+    }
+  } else {
+    current_subscript_order.reserve(lro.size() + lo.size() + ro.size());
+    current_subscript_order.insert(current_subscript_order.end(), lro.begin(), lro.end());
+    current_subscript_order.insert(current_subscript_order.end(), lo.begin(), lo.end());
+    current_subscript_order.insert(current_subscript_order.end(), ro.begin(), ro.end());
+  }
+
+  // Multiply the mutated inputs
+  auto output = MatMul<T>(left, right, allocator, tp);
+
+  // Bring the output back to the original rank (MatMul operated on 3D inputs)
+  CreateReshapedView(output, output_dims);
+
+  if (!is_final_pair) {  // This is not the final pair - so bring the axes order to what the inputs conformed to
+    if (IsTransposeRequired(output.Shape().GetDims().size(), output_permutation)) {
+      output = Transpose(output, output_permutation, allocator);
+    }
+  } else {  // This is the final pair - Transpose directly to the output ordering required and copy the contents to the op's output
+    FinalizeOutput<T>(output, current_subscript_order,
+                      einsum_compute_preprocessor.GetMappedSubscriptIndicesToOutputindices(), final_output,
+                      einsum_compute_preprocessor.GetOutputDims(), allocator);
+  }
+
+  return output;
+}
+
+}  // namespace EinsumOp
+
+EinsumComputePreprocessor::EinsumComputePreprocessor(EinsumEquationPreprocessor& einsum_equation_preprocessor,
+                                                     const std::vector<const Tensor*>& inputs,
```

Comment:
nit: I'd be inclined to separate the construction/initialization of the class from doing the actual work by adding a method that does the preprocessing. that method could return a status if something fails.

Reply:
Done.